### DISCUSSION TITLE
Help with PIO on RP2040

### DISCUSSION THREAD

**Discussion by jdh1955** [tabemann/zeptoforth#120] [Q&A]
Hello
I am new to Zeptoforth having been using Mecrisp on an RP2040. I am hoping to tame the RP2040 PIO but have no experience and am hoping to find some example Forth code to start from - something to flash the LED perhaps. I found this https://github.com/tabemann/zeptoforth/wiki/Programmable-IO-on-the-Raspberry-Pi-Pico-with-zeptoforth but have failed to get it running. Are there any suitable examples around? Thanks in anticipation.
jdh1955

**Comment by tabemann** [comment]
On Thu, Jul 25, 2024 at 9:48 AM jdh1955 ***@***.***> wrote:

> Hello
> I am new to Zeptoforth having been using Mecrisp on an RP2040. I am hoping
> to tame the RP2040 PIO but have no experience and am hoping to find some
> example Forth code to start from - something to flash the LED perhaps. I
> found this
> https://github.com/tabemann/zeptoforth/wiki/Programmable-IO-on-the-Raspberry-Pi-Pico-with-zeptoforth
> but have failed to get it running. Are there any suitable examples around?
> Thanks in anticipation.
> jdh1955
>

I apologize for the state of that wiki page -- it is not up to date to the
latest changes to the PIO API, and it turned out there was a major bug in
the example as written. Thanks a lot for bringing it to my attention.

A good example, which is the example in the wiki page but updated to the
latest PIO API and without the bug, can be found at
https://github.com/tabemann/zeptoforth/blob/master/test/rp2040/pio_fade_blinker.fs
. I have updated the wiki page accordingly.

Thanks,

Travis

**Comment by jdh1955** [comment]
Hello Travis
Many thanks for the quick response - that’s great and it ran immediately of course. I was directed to Zeptoforth in response to looking for PIO support after using Mecrisp forth. I must say that both of these systems are very nice but the USB support and the PIO assembler will be very helpful.
Best wishes
Jonathan

**Comment by jdh1955** [comment]
Thanks again - I now have the PIO working. FYI it is doing DMA from an ADC via GPIO.
